File: android/app/src/main/kotlin/com/typography/animations/loop/Heartbeat.kt
package com.typography.animations.loop

import com.typography.AnimatedTextState
import kotlin.math.*

fun evaluateHeartbeat(t: Float, speedFactor: Float): AnimatedTextState {
    val freq = 1.2f * speedFactor
    val localT = (t * freq) % 1.0f
    val s = if (localT < 0.2f) {
        1f + sin(localT * 5f * PI.toFloat()) * 0.2f
    } else if (localT < 0.5f) {
        val t2 = (localT - 0.2f) * (1f / 0.3f)
        1f + sin(t2 * PI.toFloat()) * 0.1f
    } else {
        1f
    }
    return AnimatedTextState(scale = s)
}

File: lib/utils/animations/loop/heartbeat.dart
import 'dart:math';
import '../base/animated_text_state.dart';

AnimatedTextState evaluateHeartbeat(double t, double speedFactor) {
  final freq = 1.2 * speedFactor;
  final localT = (t * freq) % 1.0;
  double s;
  if (localT < 0.2) {
    s = 1.0 + sin(localT * 5.0 * pi) * 0.2;
  } else if (localT < 0.5) {
    final t2 = (localT - 0.2) * (1.0 / 0.3);
    s = 1.0 + sin(t2 * pi) * 0.1;
  } else {
    s = 1.0;
  }
  return AnimatedTextState(scale: s);
}
